File: examples/cpp/sealctracetest/readtrace.cpp
#include <stdlib.h>
#include <stdio.h>
#include <stdint.h>
#include <string.h>

#include <iostream>

#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <arpa/inet.h>

#include <cjson/cJSON.h>

#include "seal/seal.h"

#define INTERACTIVE (0)

const char *TESTJSON = "testcore.scs";
const char *REG_TEST_TRACE = "testtrace.stc";
const char *ITA_TEST_TRACE = "/tmp/sealtraceif";

using namespace std;

void PrintFrame(SealFrame * frame)
{
    int i;

    printf("Frame mapping:");
    for (i = 0; i < frame->len; i++)
	{
	    if (!(i % 8))
		printf("\n[%02d]: ", i);

	    printf("%02X ", frame->buf[i]);
	}
    cout << endl;
    return;
}

void PrintComponent(SealFrameCompInst * idx)
{
    int i;

    printf("|%-12s|%-8s|%02d| = [%012lX] ",
	   idx->name, SccGetTypeStr(idx->type), idx->len, (long)idx->val);
    for (i = 0; i < (SccSizeOf(idx->type) * idx->len); i++)
	{
	    printf("%02X ", ((unsigned char *)idx->val)[i]);
	}

    if (OCTET != idx->type)
	{
	    printf("(");
	    //Translated info.
	    if (STRING == idx->type)
		{
		    printf("%s", (const char *)idx->val);
		}
	    else
		{
		    for (i = 0; i < idx->len; i++)
			{
			    switch (idx->type)
				{
				case INT32:
				    printf("%d",
					   ntohl(((const int32_t *)idx->
						  val)[i]));
				    break;

				case UINT32:
				    printf("%d",
					   ntohl(((const uint32_t *)idx->
						  val)[i]));
				    break;

				default:
				    break;
				}

			    if (idx->len - 1 > i)
				printf(",");
			}
		}
	    printf(")");
	}
    printf("\n");
    return;
}

void ReadRecord(SealCore * core)
{
    SealFrame *frame = NULL;
    SealFrameCompInst rega, regb, exeins, mode, cyclecount;
    SealTrace *trace = NULL;
    int cycle, ret;

    //Init
#if INTERACTIVE
    trace =
	StOpen(REG_TEST_TRACE, core, SEAL_TRACE_MODE_READ,
	       SEAL_TRACE_FLAG_INTERACTIVE);
#else
    trace = StOpen(REG_TEST_TRACE, core, SEAL_TRACE_MODE_READ);
#endif
    frame = SfNewFrame(core);

    //Set up index to components.
    SfFetchFrameCI(frame, &rega, "RegA");
    SfFetchFrameCI(frame, &regb, "RegB");
    SfFetchFrameCI(frame, &exeins, "ExeIns");
    SfFetchFrameCI(frame, &mode, "Mode");
    SfFetchFrameCI(frame, &cyclecount, "CycleCount");

    for (cycle = 0; 0 < (ret = StNextFrame(trace, frame)) && cycle <= 11;
	 cycle++)
	{
	    //Print each Core Component.
	    printf("===== CYCLE: %02d =====\n", cycle);
	    printf("Frame ID: %ld (Header = %d)\n", trace->offset,
		   ntohl(*frame->header));
	    PrintComponent(&rega);
	    PrintComponent(&regb);
	    PrintComponent(&exeins);
	    PrintComponent(&mode);
	    PrintComponent(&cyclecount);
	    PrintFrame(frame);
	    printf("=====************=====\n\n");
	}

    StClose(trace);
    SfDeleteFrame(frame);

    return;
}

int main(int argc, char *argv[])
{
    SealCore *core = NULL;

    if (argc < 2)
	{
	    printf("Usage: readtrace ${TRACE_FILE_PATH}\n");
	    exit(0);
	}
    REG_TEST_TRACE = argv[1];

    //Read a Core.
    core = ScLoadCore(TESTJSON);

    //Read Trace file.
    ReadRecord(core);

    ScDeleteCore(core);
    return 0;
}

File: examples/Python3/readtrace.py
#!/usr/bin/python3
# Example for reading a Trace file.

import sys
import seal

# Verbose flag.
verbose = False


def PrintComponent(comp):
    global verbose

    print("|{:<12s}|{:<8s}|{:02d}| =".format(
        comp.name, comp.type, comp.len, comp.val), end='')

    if verbose:
        for i in range(comp.Size()):
            print(" {:02X}".format(comp.raw[i]), end='')
        pass

    print(' (', end='')
    if comp.type == 'BOOL':
        for i in range(comp.len):
            if comp.val[i]:
                print('T ', end='')
                pass
            else:
                print('F ', end='')
                pass
            pass
        pass
    elif comp.type == 'STRING':
        print("{:s}".format(comp.val), end='')
        pass
    elif comp.type in {'OCTET', 'INT16', 'UINT16', 'INT32', 'UINT32'}:
        for i in range(comp.len):
            print("{:02X} ".format(comp.val[i]), end='')
            if i != len(comp.val) - 1:
                print(',', end='')
    else:
        raise ("Unknown Component type")
    print(')', end='')

    print('')
    return


def main(argc, argv):
    global verbose

    # Cmd args check.
    if argc < 3 or '-h' in argv or '--help' in argv:
        print(
            "Usage: {}  {{Core Specification}} {{Trace file or FIFO}} [-v|--verbose] [-h|--help]".format(argv[0]))
        return 0

    # Set verbose flag.
    if '-v' in argv or '--verbose' in argv:
        verbose = True
        pass

    # Read Core specification.
    core = seal.Core.Load(argv[1])
    st = seal.Trace(core)

    # Open Trace file.
    st.Open(argv[2])

    count = 0
    while True:
        # Print each Frame.
        frame = st.NextFrame()
        if not frame:
            break
        print("======= FRAME {:02} ======= (Version = {})".format(
            count, frame.version))
        for i in frame.components:
            PrintComponent(frame.components[i])
            pass
        count += 1
        print("========********========\n")

    return 0


if __name__ == "__main__":
    exit(main(len(sys.argv), sys.argv))
